third_party/txt/src/txt/paragraph.h

```diff
@@ -135,374 +97,42 @@ class Paragraph {
-    bool is_rtl() const { return direction_ == TextDirection::rtl; }
-    // Tracks if the run represents trailing whitespace.
-    bool is_ghost() const { return is_ghost_; }
-    bool is_placeholder_run() const { return placeholder_run_ != nullptr; }
-
-   private:
-    size_t start_, end_;
-    TextDirection direction_;
-    const TextStyle* style_;
-    bool is_ghost_;
-    PlaceholderRun* placeholder_run_ = nullptr;
-  };
-
-  struct GlyphPosition {
-    Range<size_t> code_units;
-    Range<double> x_pos;
-
-    GlyphPosition(double x_start,
-                  double x_advance,
-                  size_t code_unit_index,
-                  size_t code_unit_width);
-
-    void Shift(double delta);
-  };
-
-  struct GlyphLine {
-    // Glyph positions sorted by x coordinate.
-    const std::vector<GlyphPosition> positions;
-    const size_t total_code_units;
-
-    GlyphLine(std::vector<GlyphPosition>&& p, size_t tcu);
-  };
-
-  struct CodeUnitRun {
-    // Glyph positions sorted by code unit index.
-    std::vector<GlyphPosition> positions;
-    Range<size_t> code_units;
-    Range<double> x_pos;
-    size_t line_number;
-    SkFontMetrics font_metrics;
-    TextDirection direction;
-    const PlaceholderRun* placeholder_run;
-
-    CodeUnitRun(std::vector<GlyphPosition>&& p,
-                Range<size_t> cu,
-                Range<double> x,
-                size_t line,
-                const SkFontMetrics& metrics,
-                TextDirection dir,
-                const PlaceholderRun* placeholder);
-
-    void Shift(double delta);
-  };
-
-  // Holds the laid out x positions of each glyph.
-  std::vector<GlyphLine> glyph_lines_;
-
-  // Holds the positions of each range of code units in the text.
-  // Sorted in code unit index order.
-  std::vector<CodeUnitRun> code_unit_runs_;
-  // Holds the positions of the inline placeholders.
-  std::vector<CodeUnitRun> inline_placeholder_code_unit_runs_;
-
-  // The max width of the paragraph as provided in the most recent Layout()
-  // call.
-  double width_ = -1.0f;
-  double longest_line_ = -1.0f;
-  double max_intrinsic_width_ = 0;
-  double min_intrinsic_width_ = 0;
-  double alphabetic_baseline_ = FLT_MAX;
-  double ideographic_baseline_ = FLT_MAX;
-
-  bool needs_layout_ = true;
-
-  struct WaveCoordinates {
-    double x_start;
-    double y_start;
-    double x_end;
-    double y_end;
-
-    WaveCoordinates(double x_s, double y_s, double x_e, double y_e)
-        : x_start(x_s), y_start(y_s), x_end(x_e), y_end(y_e) {}
-  };
-
-  // Passes in the text and Styled Runs. text_ and runs_ will later be passed
-  // into breaker_ in InitBreaker(), which is called in Layout().
-  void SetText(std::vector<uint16_t> text, StyledRuns runs);
-
-  void SetParagraphStyle(const ParagraphStyle& style);
+  virtual ~Paragraph() = default;
 
-  void SetFontCollection(std::shared_ptr<FontCollection> font_collection);
+  virtual double GetMaxWidth() = 0;
```

Comment:
I would prefer if the documentation/comments regarding what each of these methods do exist in the interface definition. Without an implementation to directly back it up, these methods can quickly get ambiguous to people trying to learn this code.

Also, pointing out the existing implementations of this interface can be very helpful.

Reply:
done